### PR TITLE
[config][generic-update] Adding apply-patch, rollback, checkpoints commands

### PR DESCRIPTION
**Pull request by ghooo**
#### What I did
Adding apply-patch, rollback, replace, checkpoint, delete-checkpoint, list-checkpoints functionality.

#### How I did it
This PR is implementing the first step in in README.md in the design document: https://github.com/Azure/SONiC/pull/736  

#### How to verify it
Using unit-tests

#### New command output (if the output of a command-line utility has changed)

```sh
admin@sonic:~$ sudo config apply-patch --help
Usage: config apply-patch [OPTIONS] PATCH_FILE_PATH

  Apply given patch of updates to Config. A patch is a JsonPatch which
  follows rfc6902. This command can be used do partial updates to the config
  with minimum disruption to running processes. It allows addition as well
  as deletion of configs. The patch file represents a diff of ConfigDb(ABNF)
  format or SonicYang format.

  <patch-file-path>: Path to the patch file on the file-system.

Options:
  -f, --format [CONFIGDB|SONICYANG]
                                  format of config of the patch is either
                                  ConfigDb(ABNF) or SonicYang
  -d, --dry-run                   test out the command without affecting
                                  config state
  -v, --verbose                   print additional details of what the
                                  operation is doing
  -h, -?, --help                  Show this message and exit.



admin@sonic:~$ sudo config replace --help
Usage: config replace [OPTIONS] TARGET_FILE_PATH

  Replace the whole config with the specified config. The config is replaced
  with minimum disruption e.g. if ACL config is different between current
  and target config only ACL config is updated, and other config/services
  such as DHCP will not be affected. **WARNING** The target config file
  should be the whole config, not just the part intended to be updated.

  <target-file-path>: Path to the target file on the file-system.

Options:
  -f, --format [CONFIGDB|SONICYANG]
                                  format of target config is either
                                  ConfigDb(ABNF) or SonicYang
  -d, --dry-run                   test out the command without affecting
                                  config state
  -v, --verbose                   print additional details of what the
                                  operation is doing
  -h, -?, --help                  Show this message and exit.



admin@sonic:~$ sudo config rollback --help
Usage: config rollback [OPTIONS] CHECKPOINT_NAME

  Rollback the whole config to the specified checkpoint. The config is
  rolled back with minimum disruption e.g. if ACL config is different
  between current and checkpoint config only ACL config is updated, and
  other config/services such as DHCP will not be affected.

  <checkpoint-name>: The checkpoint name, use `config list-checkpoints`
  command to see available checkpoints.

Options:
  -d, --dry-run   test out the command without affecting config state
  -v, --verbose   print additional details of what the operation is doing
  -?, -h, --help  Show this message and exit.



admin@sonic:~$ sudo config checkpoint --help
Usage: config checkpoint [OPTIONS] CHECKPOINT_NAME

  Take a checkpoint of the whole current config with the specified
  checkpoint name.

  <checkpoint-name>: The checkpoint name, use `config list-checkpoints`
  command to see available checkpoints.

Options:
  -v, --verbose   print additional details of what the operation is doing
  -h, -?, --help  Show this message and exit.



admin@sonic:~$ sudo config delete-checkpoint --help
Usage: config delete-checkpoint [OPTIONS] CHECKPOINT_NAME

  Delete a checkpoint with the specified checkpoint name.

  <checkpoint-name>: The checkpoint name, use `config list-checkpoints`
  command to see available checkpoints.

Options:
  -v, --verbose   print additional details of what the operation is doing
  -h, -?, --help  Show this message and exit.



admin@sonic:~$ sudo config list-checkpoints --help
Usage: config list-checkpoints [OPTIONS]

  List the config checkpoints available.

Options:
  -v, --verbose   print additional details of what the operation is doing
  -?, -h, --help  Show this message and exit.
```

### CODE REVIEW

File: config/generic_update.py
Comment (qiluo-msft) [inline config/generic_update.py:314]: >) [](start = 39, length = 1)

You miss `:` at the end of line.
It was detected by lgtm linter. Please treat every lgtm alert seriosly.

File: tests/generic_update_test.py
Comment (qiluo-msft) [inline tests/generic_update_test.py:9]: >Test [](start = 6, length = 4)

Let's add a simple test case which can catch your syntax error at least.

File: config/generic_update.py
Comment (qiluo-msft) [inline config/generic_update.py:4]: >/usr/local/bin/sonic-cfggen [](start = 29, length = 27)

Is it possible not to assume the installation path? I know it may be tricky to import a script.

File: config/generic_update.py
Comment (qiluo-msft) [inline config/generic_update.py:17]: >ConfigLocker [](start = 6, length = 12)

Do you mean `ConfigLock` ? Locker means something different.

File: config/generic_update.py
Comment (qiluo-msft) [inline config/generic_update.py:50]: >config_as_yang [](start = 8, length = 14)

What is this? seems syntax error.

File: config/generic_update.py
Comment (qiluo-msft) [inline config/generic_update.py:59]: >Exception [](start = 15, length = 9)

Do not catch general exception type unless you have a good reason.

File: config/generic_update.py
Comment (qiluo-msft) [inline config/generic_update.py:63]: >data = dict() [](start = 28, length = 13)

We prefer use 2 lines.

File: config/generic_update.py
Comment (qiluo-msft) [inline config/generic_update.py:52]: >debug [](start = 50, length = 5)

Where is `debug` defined? syntax error?

File: config/generic_update.py
Comment (qiluo-msft) [inline config/generic_update.py:90]: >: [](start = 150, length = 1)

Too long line

File: config/generic_update.py
Comment (qiluo-msft) [inline config/generic_update.py:96]: >format == "ConfigDB": [](start = 11, length = 21)

Is it better to use an enum?

File: config/generic_update.py
Comment (qiluo-msft) [inline config/generic_update.py:101]: >AttributeError [](start = 18, length = 14)

ValueError will be better.

File: config/generic_update.py
Comment (qiluo-msft) [inline config/generic_update.py:116]: >Exception [](start = 18, length = 9)

Not to throw general exception type. This looks a normal logic, not like a literal 'exception'. How about using return value?

File: config/generic_update.py
Comment (qiluo-msft) [inline config/generic_update.py:239]: >__ [](start = 8, length = 2)

Most function names are prefix with double underscore, which makes the file very busy.
Do you have a strong reason to hide them?

File: generic_config_updater/generic_updater.py
Comment (qiluo-msft) [inline generic_config_updater/generic_updater.py:9]: >sonic_cfggen [](http://example.com/codeflow?start=13&length=12)

this is a new dependency. please add it in install_requires in setup.py

File: generic_config_updater/generic_updater.py
Comment (qiluo-msft) [inline generic_config_updater/generic_updater.py:29]: >PatchOrderer [](http://example.com/codeflow?start=6&length=12)

`Orderer` looks weird. How about `Sorter` ?

File: generic_config_updater/generic_updater.py
Comment (qiluo-msft) [inline generic_config_updater/generic_updater.py:73]: >.keys() [](http://example.com/codeflow?start=44&length=7)

If you are using python3, no need to add `.keys()`

File: generic_config_updater/generic_updater.py
Comment (qiluo-msft) [inline generic_config_updater/generic_updater.py:76]: >container [](http://example.com/codeflow?start=24&length=9)

`container` is from function input. So it may not contain a `:`. Then this line will throw.

File: generic_config_updater/generic_updater.py
Comment (qiluo-msft) [inline generic_config_updater/generic_updater.py:97]: >Exception [](http://example.com/codeflow?start=15&length=9)

No generic exception until you have a good reason.

File: generic_config_updater/generic_updater.py
Comment (qiluo-msft) [inline generic_config_updater/generic_updater.py:117]: >**db_kwargs [](http://example.com/codeflow?start=38&length=11)

If you don't have specific arg, you can remove the arg.

File: generic_config_updater/generic_updater.py
Comment (qiluo-msft) [inline generic_config_updater/generic_updater.py:133]: >split [](http://example.com/codeflow?start=39&length=5)

If no `/` in the string, index operator will throw

File: generic_config_updater/generic_updater.py
Comment (qiluo-msft) [inline generic_config_updater/generic_updater.py:138]: >Exception [](http://example.com/codeflow?start=22&length=9)

Do not throw generic exception type.

File: generic_config_updater/generic_updater.py
Comment (qiluo-msft) [inline generic_config_updater/generic_updater.py:159]: >Exception [](http://example.com/codeflow?start=18&length=9)

Do not throw generic exception type.

File: generic_config_updater/generic_updater.py
Comment (qiluo-msft) [inline generic_config_updater/generic_updater.py:179]: This identation looks weird. Try some python code formatter or linter, for example [http://pep8online.com/](http://pep8online.com/)

File: generic_config_updater/generic_updater.py
Comment (qiluo-msft) [inline generic_config_updater/generic_updater.py:186]: >1.  [](http://example.com/codeflow?start=10&length=3)

Normally we don't use numbered list in code comment. It's hard to maintain if you can code, add complex control logic. It is a subject opionion.

File: generic_config_updater/generic_updater.py
Comment (qiluo-msft) [inline generic_config_updater/generic_updater.py:209]: >PatchApplier() [](http://example.com/codeflow?start=39&length=14)

Do not use complex mutable object as default value. ref: [https://docs.quantifiedcode.com/python-anti-patterns/correctness/mutable_default_value_as_argument.html](https://docs.quantifiedcode.com/python-anti-patterns/correctness/mutable_default_value_as_argument.html)

File: generic_config_updater/generic_updater.py
Comment (qiluo-msft) [inline generic_config_updater/generic_updater.py:216]: >Exception [](http://example.com/codeflow?start=18&length=9)

Do not throw generic exception type.

File: generic_config_updater/generic_updater.py
Comment (qiluo-msft) [inline generic_config_updater/generic_updater.py:225]: >Exception [](http://example.com/codeflow?start=18&length=9)

Do not throw generic exception type.

File: generic_config_updater/generic_updater.py
Comment (qiluo-msft) [inline generic_config_updater/generic_updater.py:239]: >Exception [](http://example.com/codeflow?start=18&length=9)

Do not throw generic exception type.

File: generic_config_updater/generic_updater.py
Comment (qiluo-msft) [inline generic_config_updater/generic_updater.py:262]: >Exception [](http://example.com/codeflow?start=18&length=9)

Do not throw generic exception type.

File: generic_config_updater/generic_updater.py
Comment (qiluo-msft) [inline generic_config_updater/generic_updater.py:283]: >[f[:-len(CHECKPOINT_EXT)] for f in os.listdir(self.checkpoints_dir) if f.endswith(CHECKPOINT_EXT)] [](http://example.com/codeflow?start=15&length=98)

This line is complex. Better to use a normal loop and add a block comment to explain your method, even better with an example.

File: generic_config_updater/generic_updater.py
Comment (qiluo-msft) [inline generic_config_updater/generic_updater.py:346]: Weird indentation

File: tests/generic_config_updater/generic_updater_test.py
Comment (qiluo-msft) [inline tests/generic_config_updater/generic_updater_test.py:9]: >generic_config_updater [](http://example.com/codeflow?start=54&length=22)

It's a module, we'd better import a module instead of loading a single python file

File: tests/generic_config_updater/generic_updater_test.py
Comment (qiluo-msft) [inline tests/generic_config_updater/generic_updater_test.py:1059]: Weird indentation

**Review by qiluo-msft — CHANGES_REQUESTED** [review]
As comments

File: generic_config_updater/generic_updater.py
Comment (qiluo-msft) [inline generic_config_updater/generic_updater.py:1]: How do you integrate the new module with `config` ? Add the test for that integration

File: generic_config_updater/generic_updater.py
Comment (praveen-li) [inline generic_config_updater/generic_updater.py:21]: Please see if this helps:
https://github.com/Azure/sonic-utilities/pull/857/files

Note: We may have a few bug fixes on top of it.

File: generic_config_updater/generic_updater.py
Comment (praveen-li) [inline generic_config_updater/generic_updater.py:63]: using a Private function, We can expose it. By the way, why we need to directly work with sonic_yang.json

**Review by praveen-li — COMMENTED** [review]
I will review the sonic_yang part more tomorrow. Thx.

File: generic_config_updater/generic_updater.py
Comment (ghooo) [reply generic_config_updater/generic_updater.py:63]: Our teams is working on interfacing with SONiC switches using SONiC YANG instead of ConfigDB.

In the current PR I am also using JsonPatch of YANG configs and passing it to the patch ordering mechanism. I am going to switch back to JsonPatch of ConfigDB. I would hold on reviewing this PR until it is updated. Will ping you then.

If you have a moment. Can you approve the design PR if no concerns? https://github.com/Azure/SONiC/pull/736

File: generic_config_updater/generic_updater.py
Comment (ghooo) [reply generic_config_updater/generic_updater.py:21]: Yeah, this was merged then reverted. I hope we have it merged back to azure soon.

File: generic_config_updater/generic_updater.py
Comment (ghooo) [reply generic_config_updater/generic_updater.py:133]: looks OK
```python
operation = {}
operation['path']=''
tokens = operation['path'].split('/')[1:]
print(tokens)
[]
```

File: config/generic_update.py
Comment (ghooo) [reply config/generic_update.py]: will implement locking in the future once the #857 is merged back to azure repo

File: generic_config_updater/generic_updater.py
Comment (ghooo) [reply generic_config_updater/generic_updater.py:10]: let me try this idea

File: generic_config_updater/generic_updater.py
Comment (ghooo) [reply generic_config_updater/generic_updater.py:1]: done

File: generic_config_updater/generic_updater.py
Comment (ghooo) [reply generic_config_updater/generic_updater.py:10]: Was not able to include in install_requires: tried `sonic_cfggen` and `sonic-cfggen`.

Also when checking the code behind `show runningconfiguration all` it is implemeted as:
```python
@runningconfiguration.command()
@click.option('--verbose', is_flag=True, help="Enable verbose output")
def all(verbose):
    """Show full running configuration"""
    cmd = "sonic-cfggen -d --print-data"
    run_command(cmd, display_cmd=verbose)
```

Which means they just depend on the system to have, and we here are doing the same.

We can either leave it as as, or just rely on the command from cli

Comment (qiluo-msft) [reply generic_config_updater/generic_updater.py]: The name is `sonic-config-engine`. Currently it is `tests_require`. We should remove it, and add as `install_requires`.

File: .azure-pipelines/docker-sonic-vs/Dockerfile
Comment (ghooo) [inline .azure-pipelines/docker-sonic-vs/Dockerfile:7]: `docker-sonic-vs` is retrieved, then the new sonic-utilities package is installed with most recent modifications. Problem happens if the new package has new external libs or one of the libs is updated as using the command `pip3 install` with `--deps` just leave the deps unchanged.

I tried different ideas:
`pip3 install /wheels/sonic_utilities-1.2-py3-none-any.whl`
does not do anything because the package is already installed in the `docker-sonic-vs`

`pip3 install --force-reinstall /wheels/sonic_utilities-1.2-py3-none-any.whl`
does not work because it also forces the deps to be updated but some of the deps such as `sonic-py-common` are not available in public pip repo and are installed with the sonic image in `docker-sonic-vs`

`pip3 install --update /wheels/sonic_utilities-1.2-py3-none-any.whl`
does not update sonic-utilities package and only updates its dependencies

The proposed soln achieved the following:
- Force install sonic-utilities package
- Adding missing dependencies
- Upgrading out-dated dependencies
- Leaving dependencies provided by the sonic-build process unchanged.

File: generic_config_updater/generic_updater.py
Comment (qiluo-msft) [reply generic_config_updater/generic_updater.py]: That for testing! you are right.

File: generic_config_updater/generic_updater.py
Comment (ghooo) [reply generic_config_updater/generic_updater.py]: oh, let me try that

File: generic_config_updater/generic_updater.py
Comment (qiluo-msft) [reply generic_config_updater/generic_updater.py:1]: I see you modified `config` code, but only integrated with commented code.

File: generic_config_updater/generic_updater.py
Comment (ghooo) [reply generic_config_updater/generic_updater.py]: updated locally

File: generic_config_updater/generic_updater.py
Comment (ghooo) [reply generic_config_updater/generic_updater.py:1]: I might have an earlier commit that had the code commented, but that is un-commented now.

File: config/generic_update.py
Comment (ghooo) [reply config/generic_update.py]: Code updated to catch specific exceptions,

File: config/generic_update.py
Comment (ghooo) [reply config/generic_update.py]: created custom exception `GenericConfigUpdaterError`

File: config/generic_update.py
Comment (ghooo) [reply config/generic_update.py]: Got rid of `load_source` altogether.

File: generic_config_updater/generic_updater.py
Comment (ghooo) [reply generic_config_updater/generic_updater.py]: updated

File: config/main.py
Comment (qiluo-msft) [inline config/main.py:993]: >format [](http://example.com/codeflow?start=23&length=6)

This option is not in design doc.

File: config/main.py
Comment (qiluo-msft) [inline config/main.py:994]: >.lower() [](http://example.com/codeflow?start=49&length=8)

Just use the upcases?

File: config/main.py
Comment (qiluo-msft) [inline config/main.py:1018]: >secho [](http://example.com/codeflow?start=14&length=5)

`secho` to stderr if this is a critical error.

File: config/main.py
Comment (ghooo) [reply config/main.py:993]: This option is added to make it easy to use `SonicYang` format in the future once our internal services start interacting with SONiC using `SonicYang`.

File: config/main.py
Comment (ghooo) [reply config/main.py:994]: updated locally

File: config/main.py
Comment (ghooo) [reply config/main.py:1018]: updated locally

File: config/main.py
Comment (ghooo) [reply config/main.py]: updated

File: config/main.py
Comment (ghooo) [reply config/main.py]: updated

File: config/main.py
Comment (qiluo-msft) [reply config/main.py]: Better to update the design doc.

**Review by qiluo-msft — APPROVED** [review]

**Pull request merged**
